contracts/SaveToken.sol

```diff
@@ -186,24 +186,38 @@ contract SaveToken is ERC20Extended, Pausable {
     /// @notice This function will unbundle your SaveTokens for your underlying asset
     /// @param amount The number of SaveTokens to unbundle
     function withdrawForUnderlyingAsset(uint256 amount) external {
+        require(super.balanceOf(msg.sender) > 0, "Balance must be greater than 0");
+
+        // calculate ratio of amounts for the saveTokens that don't have a 1:1:1 mapping
+        (,,uint256 assetWithdrawAmount, uint256 insuranceWithdrawAmount) = _calculateRatioAmounts(msg.sender, amount);
+
         bytes memory signature_withdraw = abi.encodeWithSignature(
             "withdraw(uint256)",
-            amount
+            assetWithdrawAmount
         );
 
         bytes memory signature_sellInsurance = abi.encodeWithSignature(
             "sellInsurance(uint256)",
-            amount
+            insuranceWithdrawAmount
         );
 
-        uint256 underlyingForAsset = _delegatecall(StorageLib.assetAdapter(), signature_withdraw);
-        uint256 underlyingForInsurance = _delegatecall(StorageLib.insuranceAdapter(), signature_sellInsurance);
+        // initial underlying balance
+        uint256 initialUnderlyingBalance = StorageLib.underlyingInstance().balanceOf(address(this));
+
+        _delegatecall(StorageLib.assetAdapter(), signature_withdraw);
+        _delegatecall(StorageLib.insuranceAdapter(), signature_sellInsurance);
+
+        // updated underlying balance
+        uint256 updatedUnderlyingBalance = StorageLib.underlyingInstance().balanceOf(address(this));
+
+        // amount of the underlying to withdraw
+        uint256 underlyingToWithdraw = updatedUnderlyingBalance.sub(initialUnderlyingBalance);
 
         //transfer underlying to msg.sender
-        require(StorageLib.underlyingInstance().transfer(msg.sender, underlyingForAsset.add(underlyingForInsurance)));
+        require(StorageLib.underlyingInstance().transfer(msg.sender, underlyingToWithdraw));
 
         // update asset and insurance token balances
-        _subtractFromBalances(msg.sender, amount);
+        _subtractFromBalances(msg.sender, assetWithdrawAmount, insuranceWithdrawAmount);
```

Comment:
@dougiebuckets should we update the balances before the transfer?

Reply:
I went ahead and did the same for the `mint` function. I.e., moved updating the balances before the `_mint` of the savetokens. Cool on your end?